plugins/catalog-backend/src/catalog/types.ts

```diff
@@ -15,15 +15,18 @@
  */
 
 import { Entity, Location } from '@backstage/catalog-model';
-import type { EntityFilters } from '../database';
+import type { EntityFilters, Transaction } from '../database';
 
 //
 // Entities
 //
 
 export type EntitiesCatalog = {
   entities(filters?: EntityFilters): Promise<Entity[]>;
-  addOrUpdateEntity(entity: Entity, locationId?: string): Promise<Entity>;
+  addOrUpdateEntity(
```

Comment:
https://github.com/spotify/backstage/pull/2617 removes this and `addEntities`, just keeping the batch version.

This layer really shouldn't care about transactions though, better just add a more explicit `dryRunEntityUpdate` kinda thing instead imo

Reply:
Yeah, as the relationship code is already open for a long time, finish that one first. I will rebase my code afterwards! :)

Reply:
I would with that HigherOrderOperations would have no access to the database class, but that is not possible as the transaction has to contain changes to both the EntitiesCatalog and LocationsCatalog